### ISSUE TITLE
[mount] consistency issue with large `.git/packed-refs` files

### ISSUE THREAD

**Issue by PeterCxy** [seaweedfs/seaweedfs#2305]
**Describe the bug**

This is a weird bug and I have not been able to reproduce it outside of my specific setup, so I am going to describe exactly what I did that triggered this bug.

Basically, I am trying to host a Gitea instance (a self-hosted Git service) with seaweedfs FUSE as its backing storage. Everything seemed to be working fine, except that some repositories seem to be consistently corrupted on the FUSE mount. Any operation such as pulling and pushing would cause the service to throw out errors.

Looking deeper, I noticed that all of these broken repositories are large repositories with a lot of tags (such as my mirrors of https://github.com/mastodon/mastodon, https://github.com/matrix-org/synapse, and the Linux kernel itself). Every time this happens, it seems that it is always the `packed-refs` file of the git repository that gets corrupted -- for some reason, when it is read by some git process, an EOF is inserted into somewhere in the middle of the file, causing git's parser to fail. More strangely, this *only* happens when I try to modify the repository by `git pull` through Gitea's `synchronize remote repository` feature or push to the repository or doing anything that reads the `packed-refs` file through Git itself -- when I simply `cat` the file, it is completely uncorrupted, but when I trigger a `git pull`, and attempt to `cat` the file immediately, it shows the corrupted content.

I believe it might be related to some specific file access pattern used by Git, but I could not dig deeper into the issue right now.

**System Setup**

- SeaweedFS 2.65 (all nodes in the cluster, including the client)
- ArchLinux
- SeaweedFS local cache is enabled (but it seems to happen even if the cache is disabled -- maybe something related to metadata cache?)

**Screenshots**

For comparison, here is a pair of screenshots of the result of `cat xxx/packed-refs` right before and after a failed `git pull` operation. Notice the last line -- that's where the truncation happened

![Screenshot_20210902_210321](https://user-images.githubusercontent.com/4532423/131935518-18cff7e5-033b-4253-a445-445c763b3020.png)
![Screenshot_20210902_210955](https://user-images.githubusercontent.com/4532423/131935521-2bddcb25-6903-4bdc-ad93-fe77e6d3e47e.png)

**Additional context**

A step-by-step procedure to trigger this bug (in my setup) is:

1. Set up the SeaweedFS mountpoint and Gitea
2. Set up a mirror of a large remote repository on Gitea (such as those I mentioned)
3. Notice that the auto synchronization starts to fail after a while (usually within a few days)
4. Observe that if you trigger the `git pull` manually, the content of `packed-refs` gets truncated for no reason, but it returns to normal after a while if no `git pull` is running.

I believe this can be reproduced with just `git` and no Git hosting service like Gitea (maybe just `git clone`-ing a large repo and then doing `git pull` repeatedly on the mountpoint could trigger the bug), but again, I have not been able to dig deeper into the issue yet due to some external reasons.

**Comment by chrislusf** [comment]
this seems related to https://github.com/chrislusf/seaweedfs/issues/2169

I would suggest upgrade and see whether it can be reproduced.

**Comment by PeterCxy** [comment]
@chrislusf I'm sorry, I had a typo in my original post -- I am already on the latest version 2.65, not 2.56. This issue is present on the latest release.

**Comment by chrislusf** [comment]
I do not know how to follow step 1 and 2. Need exact details.

**Comment by PeterCxy** [comment]
@chrislusf I believe repeated `git pull` after a `git clone` (or `git clone --bare`) of said repositories might be enough to trigger the bug. But again, I have not tested it thoroughly. I'll write more about the `Gitea` setup later.

**Comment by Dando-Real-ITA** [comment]
I see the same error EOF with this setup:

- 2.70
- docker private registry with seaweedfs fuse mount as storage backend

and these steps
- create and push a docker image to the private registry
- Try to pull the image from another server
- Get an unexpected EOF error on layer download. If retry right away, the error happens again
- Wait some minutes 
- Now the image can be pulled, no more EOF errors 

Seems some async flow happens between metadata upload and data upload
Before around 2.60 I was able to pull an image 
Right after pushing

**Comment by chrislusf** [comment]
@Dando-Real-ITA could you please try latest version? there are some changes related in 2.72

**Comment by Dando-Real-ITA** [comment]
It still happens with 2.74, I will have to check with latest
```
ktadmin@KTHTTP7001:/var/www$ date
Mon Nov  8 **16:54:36** UTC 2021
ktadmin@KTHTTP7001:/var/www$ PROJECT=katapydev/web-laravel-proj-p9-amplifychurch TAG=develop DEST="/var/www/AMPLIFYCHURCH_DEMO" IMAGE=registry.katapy.io/${PROJECT}:${TAG}   && export CVSIGNORE=".env appconfig.hjson en.json sitemap.xml sitemap.xml.gz robots.txt"   && docker pull ${IMAGE}   && docker run --rm -it     -e CVSIGNORE     -v ${DEST}:/dst     ${IMAGE}
develop: Pulling from katapydev/web-laravel-proj-p9-amplifychurch
a0d0a0d46f8b: Already exists
43ab716a85ca: Already exists
f733e546095b: Already exists
80a29fc6c9e0: Pulling fs layer
error pulling image configuration: unexpected EOF
```

```
ktadmin@KTHTTP7001:/var/www$ date
Mon Nov  8 **16:55:36** UTC 2021
ktadmin@KTHTTP7001:/var/www$ PROJECT=katapydev/web-laravel-proj-p9-amplifychurch TAG=develop DEST="/var/www/AMPLIFYCHURCH_DEMO" IMAGE=registry.katapy.io/${PROJECT}:${TAG}   && export CVSIGNORE=".env appconfig.hjson en.json sitemap.xml sitemap.xml.gz robots.txt"   && docker pull ${IMAGE}   && docker run --rm -it     -e CVSIGNORE     -v ${DEST}:/dst     ${IMAGE}
develop: Pulling from katapydev/web-laravel-proj-p9-amplifychurch
a0d0a0d46f8b: Already exists
43ab716a85ca: Already exists
f733e546095b: Already exists
80a29fc6c9e0: Pull complete
Digest: sha256:9e557155d106468126e0c18fcca36490caef8d8bf76ce4a006c9d793505eb0a6
Status: Downloaded newer image for registry.katapy.io/katapydev/web-laravel-proj-p9-amplifychurch:develop
registry.katapy.io/katapydev/web-laravel-proj-p9-amplifychurch:develop
```

**Comment by Dando-Real-ITA** [comment]
I loaded the latest code from 2.78 and I still experience the EOF problem for a couple of mins
I added a pull request #2451 that helps with some issues I had with mount, like infinite reconnection when reading metadata log, but does not fix this temporary EOF problem.

**Comment by chrislusf** [comment]
Please help to verify whether this is fixed now in 3.20 or later. My local tests seem fine.

**Comment by PeterCxy** [comment]
I'm actually not sure whether this still happens or not, as I have been using JuiceFS on top of SeaweedFS as the POSIX fs implementation for a long time (instead of SeaweedFS Mount). Their POSIX implementation seems more complete, though it has to go through the S3 gateway...
